### ISSUE TITLE
[Question] Caching strategy for dependencies with Docker

### ISSUE THREAD

**Issue by aga5tya** [lerna/lerna#858]
I am trying to implement caching of node_modules on lerna based app with docker deployments for faster builds. I used a similar technique for simple apps detailed [here](http://bitjudo.com/blog/2014/03/13/building-efficient-dockerfiles-node-dot-js/), but was curious if people were using some patterns that helps under lerna based structure. Has anyone tried with some methods with lerna in specific?

**Comment by danielo515** [comment]
You should explain what are your expectations and your target architecture. 
I have several lerna mono repos, where just some packages will become docker images. So I have a docker image on each package that needs it. Then I just do `lerna run build` to build all the packages that has a docker image.

Is your architecture different?

**Comment by aga5tya** [comment]
@danielo515 that sounds similar to what i want to achieve. Could you please point me onto some example repo? would help us alot.

Ideally we need the node_modules deps to be part of a docker image and skip npm install (unless something gets changed in package.json) during re-building/re-deploying at production, for a docker image that has app code/package. Skipping npm install for deps and caching on repeat builds saves us around 5 mins. 

Right now i'm hoisting all dependencies using bootstrap command so i have a single node_modules directory at lerna root that packages utilize., But unclear as to how do i map these to docker images so they talk to each other on production environment (something linking docker volumes/images)?

**Comment by evocateur** [comment]
I have no experience with Docker, so I'm afraid I can't offer any insight.

**Comment by danielo515** [comment]
Hello @aga5tya 

Sadly I don't have any example repository, but I can provide you an outline


```sh
lerna_repo
|- package.json <- here is a global build command with a previous lerna publish
|- lerna.json
|- packages
  |- web-service
    |- package.json <- Here is the build command
    |- Dockerfile <- Here I perform a npm install
```

While in development I rely on lerna linking packages using symbolic links. Then, when I build any docker image, first I publish all the packages, and then those packages are installed inside the dockerfile like any other external package. For this task we have a private npm repository using verdaccio (the best fork of sinnopia).

This should work for your packages like any other external node.js package, and the same caching techniques should apply

Hope that works for you too.

**Comment by nkbt** [comment]
To avoid publishing packages so we can depeloy and test dev version we use this solution at the moment:

1. copy project to `build`
2. copy `node_modules` to `build` (note, that symlinks are broken at this point)
3. remove all obvious top level symlinked dev dependencies
4. for each symlinked dependency do `npm install ${symlinkedDir}`

As a result we have fully up to date (current dev/branch) dependency tree without any duplicates, so docker image is as small as when we install published dependencies.


Here is a script if you are keen to see tech details

ps: `bash` and `rsync` are just convenience wrappers around `child_process.execSync`


```js
const getSymlinks = app => {
  let dirs = [];
  try {
    dirs = fs.readdirSync(`${app}/node_modules/@private_scope/`);
  } catch (err) {
    // whatever
  }

  return dirs
    .map(dir => `${app}/node_modules/@private_scope/${dir}`)
    .map(dir => ({dir, stat: fs.lstatSync(dir)}))
    .filter(({stat}) => stat.isSymbolicLink())
    .map(({dir, stat}) => ({
      dir,
      stat,
      pkg: path.normalize(`${CWD}/node_modules/@private_scope/${fs.readlinkSync(dir)}`)
    }));
};


const getSymlinkedDirs = (app, known = []) =>
  known.includes(app) ?
    known :
    getSymlinks(app).reduce((result, {pkg}) => getSymlinkedDirs(pkg, result.concat([pkg])), known);



// For tagged release do normal `npm install --production` since everything is tagged
if (isRelease) {
  bash('npm install --production', {cwd: `${BUILD_DIR}`});
  return;
}

rsync('node_modules', `${BUILD_DIR}/`);
bash(`rm -rf ${BUILD_DIR}/node_modules/.bin`);
bash(`rm -rf ${BUILD_DIR}/node_modules/.cache`);

const {devDependencies = {}} = require(`${BUILD_DIR}/package.json`);
// Manually remove all dev deps, since prune does not pick all of them up for some reason
Object.keys(devDependencies)
  .forEach(dep => bash(`rm -rf ${BUILD_DIR}/node_modules/${dep}`));

bash('npm prune --production', {cwd: `${BUILD_DIR}`});

const symlinks = getSymlinks(BUILD_DIR);
const allPackages = getSymlinkedDirs(BUILD_DIR);

symlinks.forEach(({dir}) => bash(`rm ${dir}`, {cwd: BUILD_DIR}));
allPackages.forEach(pkg => bash(`npm install ${pkg}`, {cwd: BUILD_DIR}));
```


^^ this works quite well for us for last several months.


Still working out caching strategy tho for those dev deps to make it fast

**Comment by danielo515** [comment]
@nkbt How is your build process different or which advantages does it have over the one that I proposed  (Just making a publish and npm install in the docker container)?

**Comment by nkbt** [comment]
Well, you don't want to publish on every commit :). I was talking about building docker image on every commit with all sumlinked deps properly resolved to the latest local dev version, not published one.

**Comment by CapitanRedBeard** [comment]
@nkbt Thanks for the script. Publishing is not ideal when dealing with private repos. Care to share what your Dockerfile looks like?

**Comment by danielo515** [comment]
Hello @nkbt 

> Well, you don't want to publish on every commit :). I was talking about building docker image on every commit with all sumlinked deps properly resolved to the latest local dev version, not published one.

Your'e absolutely right about that. Sorry, I didn't saw the scenario clear at that moment.
How do you execute the script that you have posted ? Is it inside the docker file ? How do you pass the required vairables like BUILD_DIR or CWD ?

Thanks for sharing

**Comment by nkbt** [comment]
I’ll later post a better script we now use (using yarn pack) which is
million times faster. I’m on holidays now and took only my phone with me...

We are executing scripts before docker build, copying all necessary app
files and resolved deps into ‘docker/app’ folder and then doing quick and
easy ‘docker build ./docker’ which is effectively just copy files, expose
port and run app. In this case docker containers are small and don’t have
even npm. We use alpine node-base images.
Since builds are made on ci with linux there are no binary compat issues
with native node modules (if we happen to have some).
-- 
Nik Butenko
nik@butenko.me
http://github.com/nkbt

**Comment by jwickens** [comment]
@nkbt how did you get yarn pack to work? for me it just packs the current package, no dependencies

**Comment by nkbt** [comment]
@jwickens 
1. Copy all `packages/*` to `.pk`
2. Replace all internal cross-deps with packed files like (NOTE: you do it *before* packing)
  ```js
    "@monorepo/bla" => "^0.0.123"
    // becomes
    "@monorepo/bla" => "{fullpath-to-.pk}/bla.tgz"
    // relative path is essential for yarn to install deps properly
  ```
  
3. Pack every package there into the root of `.pk`
4. Move `.pk/*.tgz` and all packages `.pk/*` to `docker/app/packages`, do your normal `yarn install --production` and docker build for any of t

So as a result of these steps we are getting "production"-like build with all the internal packages installed from most recent updated versions from `tgz` to any depth (since all of them have tgz deps).

For release builds we obviously skip this.

Here is the code for `pack.js` (all other build scripts are same as usual)
```js
#!/usr/bin/env node
'use strict';


const fs = require('fs');
const util = require('util');
const path = require('path');
const glob = require('glob').sync;
const cp = require('child_process');


const {PK} = require('./lib/pkDir');
const {isRelease} = require('./lib/isRelease');


if (isRelease()) {
  console.log('Release, skipping packing');
  return;
}


const writeFile = util.promisify(fs.writeFile);
const exec = util.promisify(cp.exec);


const tgz = name => path.join(PK, `${name.replace('@myscope/', '')}.tgz`);


const pk = glob(`*/package.json`, {cwd: PK, realpath: true})
  .reduce((result, file) => {
    const json = require(file);

    return {
      ...result,
      [json.name]: {
        dir: path.dirname(file),
        file,
        tgz: tgz(json.name),
        json
      }
    };
  }, {});


// Replace all REEF deps with tgz files and pack
Promise.all(Object.keys(pk).map(k => {
  const {dir, file, json} = pk[k];
  const {dependencies = {}} = json;

  const newDeps = Object.keys(dependencies)
    .filter(name => name.startsWith('@myscope') && name in pk)
    .reduce((result, name) => ({...result, [name]: pk[name].tgz}), dependencies);

  return Promise.resolve()
    .then(() => writeFile(file, JSON.stringify({...json, dependencies: newDeps}, null, 2), 'utf-8'))
    .then(() => exec(`yarn pack --filename ${tgz(json.name)}`, {cwd: dir}));
}));
```

**Comment by jwickens** [comment]
Nice, that seems pretty neat. What I figured out yesterday was that i could use this package: https://www.npmjs.com/package/bundle-dependencies
like:
```
lerna exec --since -- yarn bundle-dependencies update
```

And then that will add all dependencies to bundled dependencies in each repo. And then using npm (not yarn which seems broken in the pack regards)

I do npm pack. With the bundled dependencies to follows the sym links and each "package" is self containing at that point. I then immediately untar it and docker build using that as my context.
```
yarn build && \
npm pack && \
tar -xvzf *.tgz && \
docker login -u gitlab-ci-token -p $CI_BUILD_TOKEN registry.gitlab.com && \
docker build -t $IMAGE_NAME ./package/ && \
```

**Comment by nkbt** [comment]
`yarn` pack works perfectly as far as I can tell (at least now in 1.x). What I love about it that if it is listed in package.json - `tgz` gets into `yarn.lock` file too. So you can keep enjoying your deterministic builds.

We actually pack some libraries and add `tgz` to currently worked on project root (if it is not part of monorepo), installing it with relative path like `yarn add ./bla.tgz`. So you have fully working `yarn.lock` + `package.json` and PR with yet unpublished dev version of a library.

When all tested and before PR is closed - library gets proper published and PR is updated with real version, and `tgz` cleaned up from history.

It's almost like `npm link` but that can be built on CI and tested in staging environment before doing actual publish.

PS: `npm pack` should work the same way. And you should be able to do `npm install xxxx.tgz` just fine too as far as I know. Haven't really tried that but when I last checked npm 5 docs it looked like working solution in npm-land too.

**Comment by sgendler-stem** [comment]
Can someone clarify the prior 2 comments?

I'm trying to build a deployable docker container, so I need to copy the contents of a deployable tgz into my container when it is building.  When I use ```yarn pack``` I get no node_modules at all (which seems to be a bug in yarn that has a recently submitted pull request awaiting final approval), but when I use ```npm pack``` I only get the node_modules that are in the private package workspace, none of the packages that were installed to {monorepo_root}/node_modules, let alone any of the symlinked packages from other workspaces.  When I run a patched version of yarn, I get the same result (but it takes 3 times as long to produce it).  

I've spent hours trying to find an npm, yarn, or lerna compatible way to flatten out the node_modules structure, but nothing seems to work.  I installed bundle-dependencies and ensured that bundleDependencies is updated in the private package, but that doesn't change the behaviour of either yarn pack or npm pack.  I can manually copy node_modules with a flag to follow symlinks, but that seems like such a broken way to do that, requiring that I either parse bundleDependencies from package.json and only copy the listed packages or else I end up with packages that are only intended for other workspaces.  Surely one of npm, yarn, or lerna can construct a complete distributable bundle without such manual intervention?  What am I missing?  I've wasted endless hours on this problem so far.

**Comment by danielo515** [comment]
> I can manually copy node_modules with a flag to follow symlinks,

If you do that you will have only the dependencies required by your package and your dependencies. What's the problem with that ?

I've been thinking about this lately and recently I started working with docker multi-stage builds.
I think that this is a great use-case for multi stage builds, where you can install lerna globally (or have an image with it already installed) install just the production dependencies and make a lerna bootstrap. Then on the final stage you can copy just the package that contains your service, with follow symlinks activated and boom, a good docker image.

Something like this:

```
FROM myLernaImage as deps
WORKDIR /home/app
RUN npm i --production && lerna bootstrap

FROM node:8-slim
WORKDIR /home/app
COPY ./packages/service .
COPY --from=deps /home/app/node_modules node_modules
...
```

Please note that this is just an outline, I'm not taking in account all the possibilities.

**Comment by jwickens** [comment]
Hi @sgendler-stem I ended up moving away from bundleDependencies and use `yarn pack` in each dir as suggested by @nkbt. First I go through each mono-repo package and for each dependency that exists in the mono-repo I change its location to be a file which will exist on an intermediary docker image. Then I run yarn pack on each dependency. The result of each bundle I move into a common directory that has a "common-node" docker image that simply copies the packed deps into the build dir, at the file location mentioned earlier. Now each package that needs a docker continues from the common docker and can run npm install to get any mono-repo dependencies.

In summary:
1. Update mono-repo cross references to point at files that will exist in the final install environment
2. Run yarn pack and collect packed npm bundles and move into common docker
3. From common docker build each separate docker

**Comment by sgendler-stem** [comment]
You guys aren't still running into the bug in yarn that doesn't add node_modules at all?  Or you've just applied the patch that hasn't been approved yet?

@jwickens , I'm not sure I'm following your process.  

I've got the following structure

```
monorepo:
  package.json
  lerna.json
  node_modules:
    @myscope:
      symlink
      symlink
    external_dependency1:
    external_dependency2:
  packages:
    pkg1:
      package.json
      node_modules:
        external_dep1
      src
    pkg2:
      package.json
      node_modules:
        external_dep3
        external_dep4
      src
  services:
    service1:
      package.json
      src
      node_modules:
        external_dep2
        external_dep3
    service2:
      package.json
      src
```
There are no cross-dependencies between services, but each service is dependent on 1 or more packages in the packages subdir.  node_modules in each package or service only contains dependencies that have conflicts with packages that are otherwise stored at the top level in monorepo/node_modules

Here's what I believe you are suggesting - 

use something like pack.js, as provided by @nkbt, to construct a tgz for each package.  It also modifies any cross dependencies in package.json to refer to the tgz file which will be created for that directory, writing the path such that it WILL resolve correctly within the container when that package is loaded - so either a relative path from where the eventual node command will run, or an absolute path within the docker image that will be built.  For ease, I'll just write them to the top level of the monorepo for now.

So now I have:

```
monorepo:
  package1.tgz
  package2.tgz
  package3.tgz
  node_modules:
    external_dep1
    external_dep2
    ...
  services
    service1:
      package.json
      node_modules:
    service2:
      package.json
      node_modules:
```

And those tgz files all live next to the top level node_modules directory in the monorepo, which contains all of the shared dependencies.

So now I need to build the docker for each directory in monorepo/services, but that still has the following structure:

```
monorepo:
  node_modules:
    huge_list_of_shared_dependencies
    ...
  services:
    service1:
      package.json
      node_modules
        short_list_conflicted_dependencies
```

So I think I need to do the following:

copy the service OUT of the monorepo and into my docker build area, so that it doesn't find the dependencies that are up a level.
Update package.json to point to tgz files just like I did for cross dependencies earlier.
Run yarn install, creating a node_modules that has ALL dependencies except for the tgz files.
Build service.
Build docker image.

Or, as I think you might be doing, create a docker image that has no service at all, but does have all of the internal monorepo dependencies in it, as tgz files.

Then, to build each service, run a build that starts with the monorepo build, adds the service, runs yarn install to pick up all the external dependencies, then builds the service.  

Correct?

**Comment by sgendler-stem** [comment]
There is still some large piece of information that is lacking here.

yarn pack and npm pack do not ever seem to include the contents of node_modules, so I figure that must be intentional.  We can refer to the tgz file for the library content, and yarn/npm must still resolve their dependencies when running yarn install in any package that is dependent on the tgz file.  

So I built tgz files that refer to other tgz files in /tmp/packages, and then I updated one of my services to refer to monorepo packages via /tmp/packages/....tgz.  When I run yarn install, however, I just get error messages, but the file in the error message is most definitely there.

```
error An unexpected error occurred: "ENOENT: no such file or directory, chmod '/Users/[redacted]/Library/Caches/Yarn/v1/.tmp/8b9efab57482828f605df4fc4294cc03/dist/queries.js' (/tmp/packages/lib-entities.tgz)".
info If you think this is a bug, please open a bug report with the information provided in "/private/tmp/app/yarn-error.log".
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

But there's also this:

```
$ ls -la /tmp/packages/lib-entities.tgz 
-rw-r--r--  1 [redacted]  wheel  8902 Jun 18 14:05 /tmp/packages/lib-entities.tgz
```

**Comment by nkbt** [comment]
Since `yarn` fixed its way of installing deps (installing from folder does not symlink anymore, but follows the same way as you "publish"/"pack" and "install") we can do way less hacky solution now!

Here is what I do now (I tried to comment as much as possible, but feel free to ask if something is not clear enough)


## 1. Workspace info
Yarn added workspaces info yay! Let's get it and use later

```bash
yarn workspaces info --json
```

## 2. Copy

We need to copy all packages since we are going to modify their `package.json` files

```js
// `bash` is just a glorified execSync 

const PK = `${MONOREPO_ROOT}/.pk`;
bash(`rm -rf ${PK}`);
Object.keys(workspaces).forEach(key => {
  bash(`mkdir -p ${PK}/${workspaces[key].location}`);
  bash(`cp -R 
    ${MONOREPO_ROOT}/${workspaces[key].location}
    ${PK}/${path.dirname(workspaces[key].location)}/`);
});
```
Note that since I actually use multiple subfolders (`services`, `libs`, `components`, etc) and not just default `packages` - I keep same structure here in `.pk` too, so no naming collisions and eady to work with

## 3. Update `package.json`

No need to `yarn pack` anymore! Much simpler and faster solution

```js
const dir = name => `${PK}/${workspaces[name].location}`;

// Replace internal dependencies with full path to that dependency
Object.keys(workspaces).forEach(name => {
  const packageJsonPath = `${PK}/${workspaces[name].location}/package.json`;
  const packageJson = require(packageJsonPath);
  if ('dependencies' in packageJson) {
    workspaces[name].workspaceDependencies.forEach(depName => {
      if (depName in packageJson.dependencies) {
        Object.assign(packageJson.dependencies, {[depName]: dir(depName)});
      }
    });
    fs.writeFileSync(packageJsonPath, JSON.stringify(packageJson, null, 2), 'utf-8');
  }
});
```
I also only care about `deps` not dev deps, since later will use `yarn install --production` for docker.

Resulting `package.json` would look like
```json
{
  "dependencies": {
    "@myscope/config": "^1.1.1",
    "@myscope/error": "/Users/nkbt/monorepo/.pk/libs/error",
    "@myscope/log": "^2.0.2",
    "@myscope/transit": "/Users/nkbt/monorepo/.pk/libs/transit",
    "component-router": "^3.0.8",
    "compression": "^1.7.1",
    "connect-timeout": "^1.9.0"
  }
}
```
Note that some of libs under `@myscope` are not part of monorepo at all, so they were not replaced

## 4. Build

That is very much app-specific. In my case I go into each service folder and do something like:

```js
const APP_DIR = `${PK}/${workspaces[name].location}`;
const DOCKER_DIR = `${APP_DIR}/docker`;
const DOCKER_APP_DIR = `${DOCKER_DIR}/app`;

// cleanup docker app dir
bash(`rm -rf ${DOCKER_APP_DIR}`);
bash(`mkdir -p ${DOCKER_APP_DIR}`);

// copy only package/lock, install prod deps
bash(`cp -R ${APP_DIR}/package.json ${DOCKER_APP_DIR}/`);
bash(`cp -R ${MONOREPO_ROOT}/yarn.lock ${DOCKER_APP_DIR}/`);
bash('yarn install --production --pure-lockfile', {cwd: `${DOCKER_APP_DIR}`});
// remove lock, since we will never need it in the app
bash(`rm -rf ${DOCKER_APP_DIR}/yarn.lock`);


// copy rest of app dirs and files
bash(`cp -R ${APP_DIR}/src ${DOCKER_APP_DIR}/`);
bash(`cp -R ${APP_DIR}/lib ${DOCKER_APP_DIR}/`);
bash(`cp -R ${APP_DIR}/index.js ${DOCKER_APP_DIR}/`);


// copy webpack config and run client build
bash(`cp -R ${APP_DIR}/webpack.config.js ${DOCKER_APP_DIR}/`);
bash(`${require.resolve(`.bin/webpack`)}`, {env: {NODE_ENV: 'production'}, cwd: DOCKER_APP_DIR});
// cleanup webpack config since we never need in in runtime app
bash(`rm -rf ${DOCKER_APP_DIR}/webpack.config.js`);
// cleanup src files since we've built everything and never use src in runtime
bash(`rm -rf ${DOCKER_APP_DIR}/src`);


// now build docker container
// Note! We run docker build inside DOCKER_DIR,
// so we do not send full monorepo context to docker, but only small app
// That is much, much faster
bash(`docker build --tag ${DOCKER_TAG} ${DOCKER_DIR}`);


// push container
bash(`docker push ${DOCKER_TAG}`);
```

**Comment by nkbt** [comment]
I hope this little guide helps. 
It works very fast and is a reliable solution we use for our services. Works in CI on each pr/commit, no problems at all. 
In this case we are getting docker containers pushed without a need to publish libs first.
Without a need to do `lerna publish --canary`, which can be an alternative solution, but I do not like to have thousands releases for each lib, which would be the case when we do builds in CI.

The only time I use `lerna publish --canary` if some libs from monorepo are used outside and I want to do a full e2e test of that outside project with new dev versions of libs (would not be a problem of course if I just keep all services inside monorepo, hehe)

**Comment by wizardnet972** [comment]
@sv2 In your solution you keep the repo structure and it's can be kind of problem when it become building images for apps separately. 
Let's say I have apps folder and packages folder. apps use packages.
I want to build an image for only one app.
If I use lerna I should have to keep packages, apps folders, and package.json insiders so bootstrap can work. - The output from my build process can be different than my app structure for example if I use webpack with nodejs (all the files are locate in `dist` folder).

**Comment by danielo515** [comment]
I still think multi stage builds are the simplest solution

**Comment by mlegenhausen** [comment]
As @danielo515 already mentioned multi stage build work very well, but you have the problem that depending on the amount of dependencies the resulting images can become very big and cleaning up to get only the deps you need is a fiddly thing.

For static front end stuff the solution is really easy you bundle everything up and copy everything in an image with a http server.

You can use the same approach for your backend by using webpack to build a single file node application. So you don't have to copy any `node_modules` folders. webpack takes care that only the files are included that you really need. Here an example webpack configuration for your node code:

```js
const CopyPlugin = require('copy-webpack-plugin');
const path = require('path');

module.exports = {
  mode: 'development',
  devtool: 'source-map',
  target: 'node',
  entry: ['source-map-support/register', 'src/server.js'],
  resolve: {
    extensions: ['.js', '.node']
  },
  output: {
    path: path.join(__dirname, 'dist'),
    filename: 'server.js'
  },
  module: {
    rules: [
      {
        test: /\.node$/,
        loader: 'native-ext-loader'
      }
    ]
  },
  node: {
    __dirname: false
  },
  plugins: [new CopyPlugin(['src'], { ignore: ['*.js'] })]
};
```

A multi stage `Dockerfile` would look like this

```Dockerfile
FROM node:10-alpine as builder

WORKDIR /app

COPY package.json lerna.json yarn.lock /app/
COPY packages/ /app/packages/

RUN yarn --frozen-lockfile
# This command builds the server code
RUN npx lerna run build


FROM node:10-alpine as server

WORKDIR /app

COPY --from=builder /app/packages/server/dist/ /app/

ENTRYPOINT ["node", "/app/server.js"]

# Extend the file for all your docker containers and only copy the resulting bundles from builder
```

We were able to reduce the size of our images from 700mb to ~70mb

**Comment by mlegenhausen** [comment]
I improved my example multi stage configuration based on this guide: https://medium.com/@tonistiigi/advanced-multi-stage-build-patterns-6f741b852fae
The improvements allows you build a single `target` image without building all packages.

**Comment by JamesHenry** [comment]
Hi Folks 👋

In light of its age, and in order to allow our small team to focus on what matters most to Lerna v6 users in late 2022, I'm going to close this one. If you have any requests that could enhance your usage of Lerna 6, then please do let us know by opening a new discussion (for a feature) or issue (for a bug report) with as much context as possible.

You can also check out our published roadmap for Lerna v7 here: https://github.com/lerna/lerna/discussions/3410

Many thanks 🙏
